### DISCUSSION TITLE
Nesting Form Switches in HOC containing useForm causing rerenders

### DISCUSSION THREAD

**Discussion by u382514** [react-hook-form/react-hook-form#2891] [Q&A]
I have an HOC that switches out the forms depending on a particular prop passed to the HOC component. useForm is inside of the HOC component and the methods are being passed as props to the child form that is being rendered. This way I don't need to duplicate code, make dumb components dependent on the library and easier to test. It seems that when the form is switched out with a new form the form will render as many times as it takes to unregister each field from the PREVIOUS form! The weird thing is that if I have an input on the new form that matches the same name it will not fire that unregister event. 

So if Form 1 has fields:
first-name
last-name
phone

and Form 2 has fields:
first-name
address
city

If I move from Form 1 to Form 2 (causing the HOC to render the new form) 
Form 2 will render 3 times: 
1. Render the initial form
2. Render again because last-name is being unregistered in the magic
3. Render again because phone is being unregistered in the magic 

first-name is not causing a render because the field name is in both forms and the unregister isn't processed. 

I'm assuming I'm doing something wrong here that is causing all these unnecessary renders and I don't want to just kill unregister via the ShouldUnregister = false (which does solve the problem of the multiple re-renders).  

Any thoughts? 

Thanks!

**Comment by bluebill1049** [comment]
can you share a CSB? from memory we don't work well with HOC, because the nature of HOC keep mount and unmount components.

**Comment by u382514** [comment]
Yes I can! Below is a very crude example but it shows what I mean. App is loading each particular form based on a change of state. In each form, all field names are different except for field 4 (SAMENAME_field4). I've added a console.log to findRemovedFieldAndRemoveListener function that will fire when unregistering fields. Click on 'Change To Form X' to change each form. 

[https://codesandbox.io/s/react-hook-formunregister-phenomenon-vnntq](url)

The particular issue I'm having is that any time it unregisters fields it sets off a render thus my components are rendering as many times as I have fields..(sort of, see below). 

The phenomenon is that if I have fields with the same name on each of the forms, the remove listener function is not called for those. 

You should see something like this in the console when hitting the change form button: 

```
UNREGISTER FIELD: form1_field1
UNREGISTER FIELD: form1_field2
UNREGISTER FIELD: form1_field3 //missing field 4 because the names are the same in each form
UNREGISTER FIELD: form1_field5
```
At any rate, I can see why HOC's are causing an issue but it's not unusual in react development to separate components from business logic entirely and utilize dependency injection via props where necessary which is at the heart of the issue. I have a complex application that has many layers of composition where those inputs are at the end of the line and forms that are dynamically generated via meta data. Unnecessary renders are costly.  

I'm now refactoring to properly use custom hooks and hope to see if I can't solve the render problem by bringing the logic closer to the source. 

Thanks for checking it out!

**Reply by bluebill1049** [reply]
I wouldn't use HOC in this case, rather two separate hooks at each form.

you can take a look at here for an alternative pattern which prevents working in production: https://react-hook-form.com/advanced-usage
